src/server/aws_lambda.ts

```diff
@@ -0,0 +1,58 @@
+// eslint-disable-next-line import/no-unresolved
+import { APIGatewayProxyHandler, APIGatewayProxyEvent, APIGatewayProxyResult, APIGatewayProxyCallback, Context } from 'aws-lambda'
+
+type SyncHandler = (
+    event: APIGatewayProxyEvent,
+    context: Context,
+    callback: APIGatewayProxyCallback,
+) => void;
+
+export type AsyncHandler = (
+    event: APIGatewayProxyEvent,
+    context: Context,
+) => Promise<APIGatewayProxyResult>;
+
+/**
+ * async (event) => async handler
+ * async (event, context) => async handler
+ * (event, context, callback) => sync handler
+ * @param handler
+ */
+function getAsyncHandler(handler: APIGatewayProxyHandler): AsyncHandler {
+    if (handler.length > 2) {
+        return ((event, context) => {
+                return new Promise((resolve, reject) => {
+                    (handler as SyncHandler)(event, context, (error, result) => {
+                        if (error === null || error === undefined) {
+                            resolve(result)
+                        } else {
+                            reject(error)
+                        }
+                    })
+                })
+        }) as AsyncHandler
+    }
+
+    return (handler as AsyncHandler)
+}
+
+export function lambdaHandler(handler: APIGatewayProxyHandler): AsyncHandler {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const hb = this
+    const asyncHandler = getAsyncHandler(handler)
```

Comment:
The idea here is that everyone still calls `lambdaHandler`, but that function now converts the handler to async if it's the old style? If so, is this totally transparent to end-users?

Reply:
It should be transparent, but we cannot be 100% sure:
If end-users are using the result of `Honeybadger.lambdaHandler` in their test code, they will be getting unexpected errors since this function does not take a callback as a third parameter anymore and it returns a promise.

Reply:
Yup, this is my main worry. Can't we write `lambdaHandler` such that it returns something isomorphic to what was given?

```js
function lambdaHandler(handler) {
  if (handler.length == 3) {
    return (event, context, cb) => {
      try {
        handler(event, context, cb);     
      } catch (e) {
        hb.notify(e);
        cb(e);
      }
    };
  } else {
  return (event, context) => {
    return new Promise(...); 
  }
}
```

Not ideal, but this way we avoid breaking any expectations.

_Aargh, Node.js is such a mess, what with error handling split across callbacks, promises and regular exceptions.😩_

Reply:
Done. Now I have to think about the type definitions. I want to avoid `any`, but it may not be possible.

Reply:
Hmm, maybe something like this?

```typescript
function lambdaHandler<T extends SyncHandler | AsyncHandler>(handler: T): T {
  ...
}
```

Reply:
Considering this [ticket](https://github.com/honeybadger-io/honeybadger-js/issues/682#issuecomment-1013816752), it might not be enough. I will check if `@types/aws-lambda` has something generic enough to support all their handlers, otherwise I believe we'll have to go with `unknown`.

Reply:
Hmm, why wouldn't it be enough? 🤔 You could define `SuncHandler` and `AsyncHandler` either generically, or with `any`/`unknown`. _That_ could get complex, if you're trying to support all the AWS event types and return values (unless they have a parent type). But for the `lambdaHandler` function itself, it simply needs to return the same type as it gets.

Also, hehe, it's quite timely that we started working on this at this time😀. I suspect that user's timeout problems were because of the Promise-returning handler being converted into a callback handler.

Reply:
Yeah, looking at the code again, I think we just need to redefine the `event` parameter and return types of the handlers to be more generic. Hopefully the AWS event types and return types inherit from a base type, so we can use that.

Reply:
> Also, hehe, it's quite timely that we started working on this at this time😀. I suspect that user's timeout problems were because of the Promise-returning handler being converted into a callback handler.

Yes!

> Yeah, looking at the code again, I think we just need to redefine the event parameter and return types of the handlers to be more generic. Hopefully the AWS event types and return types inherit from a base type, so we can use that.

Done! I was looking to find a way so that the return type of `Hb.lambdaHandler` would be narrowed to the sync or async version based on the input handler, but I couldn't make it. I will try this again at some other point.